## 13 Days of Coding: Claw Contraption Challenge 🕹️

### Problem Overview:
The challenge involves calculating the minimum number of tokens required to win prizes from claw machines, where each machine has two buttons (A and B) with different costs and movements. The goal is to determine the fewest tokens needed to win as many prizes as possible.

### Part 1: Calculating the Fewest Tokens to Win Prizes

The task is to find the optimal combination of button presses (A and B) to reach the prize's location in a claw machine.

- **Button A** moves the claw by specific amounts on both the X and Y axes.
- **Button B** moves the claw differently.
- The objective is to find the least expensive combination of presses to match the prize location.
---

In [7]:

import math

def parse_input(file_path):
    with open(file_path, "r") as file:
        data = file.read().strip().split("\n\n")

    machines = []
    for block in data:
        lines = block.split("\n")
        button_a = tuple(map(int, lines[0].split("X+")[1].split(", Y+")))
        button_b = tuple(map(int, lines[1].split("X+")[1].split(", Y+")))
        prize = tuple(map(int, lines[2].split("X=")[1].split(", Y=")))
        machines.append((button_a, button_b, prize))

    return machines

def solve_claw_machine(button_a, button_b, prize):
    a, c = button_a
    b, d = button_b
    X, Y = prize

    min_cost = float('inf')
    best_na, best_nb = None, None

    # Iterate over possible n_A values
    for n_A in range(101):
        rem_x = X - n_A * a
        rem_y = Y - n_A * c

        if rem_x % b == 0 and rem_y % d == 0:
            n_B_x = rem_x // b
            n_B_y = rem_y // d

            if n_B_x == n_B_y and n_B_x >= 0:
                cost = 3 * n_A + n_B_x
                if cost < min_cost:
                    min_cost = cost
                    best_na, best_nb = n_A, n_B_x

    return min_cost if best_na is not None else None

def main(file_path):
    machines = parse_input(file_path)
    total_cost = 0
    prizes_won = 0

    for button_a, button_b, prize in machines:
        result = solve_claw_machine(button_a, button_b, prize)
        if result is not None:
            total_cost += result
            prizes_won += 1

    print(f"Prizes Won: {prizes_won}")
    print(f"Minimum Tokens Spent: {total_cost}")

# Example usage
main("input.txt")

Prizes Won: 178
Minimum Tokens Spent: 37297



#### Key Points:
- Each button has a defined movement on both axes.
- The prize has a specific X and Y coordinate that the claw must reach.
- You must minimize the number of tokens spent, considering each button press costs a specific amount (3 tokens for A, 1 token for B).
  
**Example Input:**
- Button A: X+94, Y+34
- Button B: X+22, Y+67
- Prize: X=8400, Y=5400

**Example Solution:**
- Use a combination of presses for A and B to reach the prize with the fewest tokens.

**Total tokens for winning all possible prizes: 37,297**

---

### Part 2: Adjusting Prize Locations with a Shift

In Part 2, the prize locations are shifted by a large amount (10^13 units) on both the X and Y axes. The challenge is to adapt the solution to account for this shift and recalculate the minimum tokens needed to win the prizes.

- The positions of all prizes are increased by 10^13 on both axes.
- The task is to find the optimal button presses to win prizes with the updated prize locations.

---


In [8]:
def extended_gcd(a, b):
    if a == 0:
        return b, 0, 1
    gcd, x1, y1 = extended_gcd(b % a, a)
    x = y1 - (b // a) * x1
    y = x1
    return gcd, x, y

def find_solution(a1, a2, b1, b2, target_x, target_y):
    print(f"\nSolving for equations:")


    # Calculate determinant
    det = a1 * b2 - a2 * b1

    if det == 0:
        print("No solution - equations are dependent")
        return None

    # Using Cramer's rule
    n = (target_x * b2 - target_y * b1) / det
    m = (a1 * target_y - a2 * target_x) / det

    # Check if solution is integer and non-negative
    if n != int(n) or m != int(m) or n < 0 or m < 0:
        print(f"No valid solution: A={n}, B={m}")
        return None

    n = int(n)
    m = int(m)

    # Verify solution
    if (a1 * n + b1 * m == target_x) and (a2 * n + b2 * m == target_y):
        print(f"Found solution: A={n}, B={m}")
        return (n, m)

    return None

def solve_claw_machines_part2(input_data):
    total_tokens = 0
    possible_prizes = 0
    OFFSET = 10000000000000

    lines = input_data.strip().split('\n')

    for i in range(0, len(lines), 4):
        if i + 2 >= len(lines):
            break


        # Parse input
        a_line = lines[i].strip()
        ax = int(a_line[a_line.find('X+')+2:a_line.find(',')])
        ay = int(a_line[a_line.find('Y+')+2:])

        b_line = lines[i+1].strip()
        bx = int(b_line[b_line.find('X+')+2:b_line.find(',')])
        by = int(b_line[b_line.find('Y+')+2:])

        p_line = lines[i+2].strip()
        px = int(p_line[p_line.find('X=')+2:p_line.find(',')]) + OFFSET
        py = int(p_line[p_line.find('Y=')+2:]) + OFFSET


        # Find solution
        solution = find_solution(ax, ay, bx, by, px, py)

        if solution:
            n, m = solution
            tokens = 3 * n + m
            total_tokens += tokens
            possible_prizes += 1
            print(f"Tokens needed: {tokens}")
        else:
            print("No solution found")

    return total_tokens, possible_prizes

# Function to read input from file and process it
def read_input_file(file_path):
    with open(file_path, 'r') as file:
        input_data = file.read()
    return input_data

# Sample usage
input_file_path = 'input.txt'  # Path to the input file
input_data = read_input_file(input_file_path)
total_tokens, possible_prizes = solve_claw_machines_part2(input_data)
print(f"Total tokens needed: {total_tokens}")



Solving for equations:
No valid solution: A=78223568103.03734, B=75700227135.6813
No solution found

Solving for equations:
No valid solution: A=121867490613.5606, B=72090628267.33162
No solution found

Solving for equations:
No valid solution: A=51369863062.69863, B=239726027436.26028
No solution found

Solving for equations:
No valid solution: A=344370860932.15234, B=92715231804.07947
No solution found

Solving for equations:
Found solution: A=133604415069, B=142317746093
Tokens needed: 543130991300

Solving for equations:
Found solution: A=145631067952, B=104752171922
Tokens needed: 541645375778

Solving for equations:
No valid solution: A=32399157625.90183, B=131216588462.70241
No solution found

Solving for equations:
Found solution: A=113065326771, B=125628140884
Tokens needed: 464824121197

Solving for equations:
No valid solution: A=67493615529.80701, B=76614374412.94308
No solution found

Solving for equations:
Found solution: A=103459973069, B=101763907738
Tokens needed: 412

#### Key Points:
- The shift introduces much larger numbers and makes the problem more complex.
- The goal remains to minimize the tokens spent by finding the optimal button presses.

**Total tokens for winning all possible prizes in Part 2: 83,197,086,729,371**